File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
# 이거는 재귀다
def solution(arr):
    answer = [0] * 2
    func(0, 0, len(arr), arr, answer)
    return answer

def func(r, c, k, arr, answer):
    if k == 1:
        answer[arr[r][c]] += 1
        return
    
    # 우선 r, c 부터 r+k, c+k까지 모두 같은 수인지 확인
    possible = True
    target = arr[r][c]
    for i in range(k):
        for j in range(k):
            if arr[r+i][c+j] != target:
                possible = False
                break
        if not possible:
            break
    
    if possible:
        answer[target] += 1
        return
    
    func(r, c, k//2, arr, answer)
    func(r+(k//2), c, k//2, arr, answer)
    func(r, c+(k//2), k//2, arr, answer)
    func(r+(k//2), c+(k//2), k//2, arr, answer)
    

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    static int[] answer;
    public int[] solution(int[][] arr) {
        answer = new int[2];
        func(0, 0, arr.length, arr);
        return answer;
    }
    static void func(int r, int c, int k, int[][] arr) {
        if (k == 1) {
            answer[arr[r][c]]++;
            return;
        }
        
        boolean possible = true;
        int target = arr[r][c];
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < k; j++) {
                if (arr[r+i][c+j] != target) {
                    possible = false;
                    break;
                }
            }
            if (!possible) break;
        }
        
        if (possible) {
            answer[target]++;
            return;
        }
        
        int tmp = (int) k/2;
        func(r, c, tmp, arr);
        func(r+tmp, c, tmp, arr);
        func(r, c+tmp, tmp, arr);
        func(r+tmp, c+tmp, tmp, arr);
        
    }
}